### PR TITLE
#12355 Issue: Register latex to Cosmology.write

### LINKED ISSUE #12355: Register ``latex`` to ``Cosmology.write``
Cosmology can now read and write to files.
It would be nice to register with ``Cosmology.write`` a  method for exporting a Cosmology to a Latex table.
There are good examples of IO with Cosmology at https://github.com/astropy/astropy/tree/main/astropy/cosmology/io
and documentation at https://docs.astropy.org/en/latest/cosmology/io.html#cosmology-io

I'm thinking the ``write_latex(...)`` method would call ``cosmology.io.table.to_table()``, format the table to e.g. make `H0` -> `$H_0 \rm{[Mpc]}$` or something and then call the `QTable.write(..., format='latex')`.

### PR DESCRIPTION
**Pull request by Mubin17**
### Description

This pull request is to address Issue #12355 

Fixes #12355 

I've added a new write_latex() method to the Cosmology package that allows you to export a Cosmology object to a Latex table. 

Test:
```python
from astropy.cosmology import Planck18
from astropy.cosmology.io import latex 

latex.write_latex(Planck18, "dd.tex", overwrite=True, latex_names=True)
```
Output:
```latex
\begin{table}
\begin{tabular}{cccccccc}
cosmology & name & $$H_0 
{[\mathrm{km\,s^{-1}\,Mpc^{-1}}]}$$ & Om0 & $$T_{0} 
{[\mathrm{K}]}$$ & Neff & $$m_{nu} 
{[\mathrm{eV}]}$$ & Ob0 \\
FlatLambdaCDM & Planck18 & 67.66 & 0.30966 & 2.7255 & 3.046 & 0.0 .. 0.06 & 0.04897 \\
\end{tabular}
\end{table}
```
Please let me know if you have any feedback or suggestions for improvements.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks! If this patch were to proceed, would need change log and test too. I'll let cosmology maintainer decide if it also need a what's new entry. But let's wait to hear back first.

**Comment by nstarman** [comment]
@Mubin17, thanks for the PR. This is a very good start!
@pllim is correct that this needs a change-log and tests. 
A few additional things

1. The module `latex` needs to be imported in [`__init__.py`](https://github.com/astropy/astropy/blob/main/astropy/cosmology/io/__init__.py
2. The write method need to be registered with `Cosmology.write` (see https://github.com/astropy/astropy/blob/b3b8295c4b0478558bd0e4c6ec28bf16b90880b8/astropy/cosmology/io/mapping.py#L236-L241 for example)

File: astropy/cosmology/io/latex.py
Comment (nstarman) [inline astropy/cosmology/io/latex.py:59]:
```suggestion
        if name in ("cosmology", "name"):
```

File: astropy/cosmology/io/latex.py
Comment (Mubin17) [inline astropy/cosmology/io/latex.py:22]: Sorry about the mess I have made with all my commits I was trying to squash all commits into one guess I didn’t do it right.
@nstarman Based on the given _FORMAT_TABLE the column headers in the output should match the LaTeX code in the _FORMAT_TABLE.  However, this is not the case and I can't seem to understand why?
For example the output produced by the code below
```python
from astropy.cosmology import Planck18
Planck18.write("example_cosmology.tex", overwrite=True, format="latex")
```
Should be
Output:
```latex
\begin{table}
\begin{tabular}{cccccccc}
cosmology & name & $$H_0{[\mathrm{km\,s^{-1}\,Mpc^{-1}}]}$$ & $$\\Omega_{m,0}$$ & $$T_{0}{[\mathrm{K}]}$$ & $$N_{eff}$$ & $$m_{nu}{[\mathrm{eV}]}$$ & $$\\Omega_{b,0}$$ \\
FlatLambdaCDM & Planck18 & 67.66 & 0.30966 & 2.7255 & 3.046 & 0.0 .. 0.06 & 0.04897 \\
\end{tabular}
\end{table}
```

Comment (nstarman) [reply astropy/cosmology/io/latex.py:22]: The issue is that `O...` do not have units, so are tripping

``if not isinstance(param, Parameter) or param.unit in (None, u.one):``

File: astropy/cosmology/io/latex.py
Comment (nstarman) [inline astropy/cosmology/io/latex.py:64]: For safety, 
```suggestion
            table.rename_column(name, _FORMAT_TABLE.get(name, name))
```
will default to the original name if it's not in the format table.

File: astropy/cosmology/io/latex.py
Comment (nstarman) [inline astropy/cosmology/io/latex.py:63]:
```suggestion
        if name not in ("cosmology", "name"):
```

File: astropy/cosmology/io/latex.py
Comment (nstarman) [reply astropy/cosmology/io/latex.py:22]: I think the if-check can be skipped.

**Comment by nstarman** [comment]
Hi @Mubin17, is there any way I can help with this PR? Thanks again for working on this PR!

**Comment by Mubin17** [comment]
@nstarman Apologies for being MIA the past couple of days, but I have updated the write_latex() method and it seems to be working fine now. Before I move on to tests I think I should add the read_latex() and latex_identify() methods to the Cosmology package as well. Please let me know if it is necessary or if I should jump on to tests. I came up with the code below by referencing the read_html_table().

```python
def read_latex(filename, index=None, *, move_to_meta=False, cosmology=None,**kwargs):
    # Check that the format is 'ascii.html' (or not specified)
    format = kwargs.pop("format", "latex")
    if format != "latex":
        raise ValueError(f"format must be 'latex', not {format}")
    
    with u.add_enabled_units(cu):
        table = QTable.read(filename, format="latex", **kwargs)

    table_columns = set(table.colnames)
    for name, latex in _FORMAT_TABLE.items():
        if latex in table_columns:
            table.rename_column(latex, name)
            
    # Build the cosmology from table, using the private backend.
    return from_table(table, index=index, move_to_meta=move_to_meta, cosmology=cosmology)
```

However, upon execution, it throws this error:
```
ValueError: '.. 0.06' did not parse as unit: Syntax error parsing unit '.. 0.06' If this is meant to be a custom unit, define it with 'u.def_unit'. To have it recognized inside a file reader or other code, enable it with 'u.add_enabled_units'. For details, see https://docs.astropy.org/en/latest/units/combining_and_defining.html```

File: astropy/cosmology/io/latex.py
Comment (nstarman) [inline astropy/cosmology/io/latex.py:63]: For this it's actually easier to keep the unit! The table will have a separate row with the unit, which is fine for now. Probably later (in a future PR) it would be good to have an option like `inline_units`, but that's a concern for later.

**Review by nstarman — COMMENTED** [review]
> Before I move on to tests I think I should add the read_latex() and latex_identify() methods

Let's keep this PR atomic, only writing for now. I have a couple of suggestions, mostly about units. I think this PR is ready for some tests, it's looking quite good!

**Comment by Mubin17** [comment]
@nstarman can you please review these tests before I commit text_latex.py
Please let me know if you have any suggestions.
```python
# Licensed under a 3-clause BSD style license - see LICENSE.rst

# THIRD PARTY
import pytest

# LOCAL
import astropy.units as u
from astropy.cosmology.io.latex import _FORMAT_TABLE, write_latex
from astropy.table import QTable
from astropy.cosmology import Planck15

class TestWriteLatex:
    def setup_class(self):
        # Create a cosmology object to use for testing
        self.cosmo = Planck15

    def test_write_latex(self, tmp_path):
        # Write the cosmology to a temporary file
        filename = tmp_path / "test_write_latex.tex"
        write_latex(self.cosmo, filename)

        # Read the table back in using QTable.read() and check that the data is correct
        table = QTable.read(filename, format="latex")
        
        for column_name in table.colnames[2:]:  
            assert column_name in _FORMAT_TABLE.values()

    def test_write_latex_bad_cls(self, tmp_path):
        # Test that passing an invalid class to write_latex() raises a TypeError
        filename = tmp_path / "test_write_latex_bad_cls.tex"
        with pytest.raises(TypeError):
            write_latex(self.cosmo, filename, cls=str)

    def test_write_latex_invalid_path(self):
        # Test that passing an invalid path to write_latex() raises a IOError
        invalid_path = ""

        with pytest.raises(IOError):
            write_latex(self.cosmo, invalid_path, overwrite=True)

    def test_write_latex_invalid_format(self, tmp_path):
        filename = tmp_path / "test_write_latex_invalid_format.tex"
        # Test using non-Latex format
        with pytest.raises(ValueError):
            write_latex(self.cosmo, filename, format="ascii.html")
```

File: astropy/cosmology/io/latex.py
Comment (nstarman) [inline astropy/cosmology/io/latex.py]: We actually need to keep this bit, just removing the `.value`.

```
cosmo_cls = type(cosmology)
    for name, col in table.columns.copy().items():
        param = getattr(cosmo_cls, name, None)
        if not isinstance(param, Parameter) or param.unit in (None, u.one):
            continue
        table[name] <<= param.unit  # Get column to correct unit
```

**Comment by nstarman** [comment]
From a glance, these tests look good. I would try to implement them in our test suite.
Also take a look at the [``test_html``](https://github.com/astropy/astropy/blob/main/astropy/cosmology/io/tests/test_html.py) file for some more tests to adapt. It's probably easiest to copy-paste-adapt from [``test_html.py``](https://github.com/astropy/astropy/blob/main/astropy/cosmology/io/tests/test_html.py).
Last, the test needs to be hooked up in [``test_connect``](https://github.com/astropy/astropy/blob/main/astropy/cosmology/tests/test_connect.py). Ping me if anything is confusing!

File: astropy/cosmology/io/tests/test_latex.py
Comment (nstarman) [inline astropy/cosmology/io/tests/test_latex.py:46]: This test is failing because file type suffixes don't need to match the format of the contained data. It's poor form but technically allowed to save html to a `.tex`.

File: astropy/cosmology/tests/test_connect.py
Comment (nstarman) [reply astropy/cosmology/tests/test_connect.py:38]: e.g.

```
    @pytest.mark.parametrize("format, metaio, has_deps", readwrite_formats)
    def test_readwrite_complete_info(self, cosmo, tmp_path, format, metaio, has_deps):
        if not has_deps:
            pytest.skip("missing a dependency")
        if (format, Cosmology) not in readwrite_registry._readers:
            pytest.xfail(f"no read method is registered for format {format!r}")
```

Comment (nstarman) [reply astropy/cosmology/tests/test_connect.py:38]: The if-statement just needs to be added to a few places in this file. Hope this helps!

**Review by nstarman — COMMENTED** [review]
The main bit of the CI is failing, for what appears to be unrelated reasons. Fixing the changeling should also hopefully trigger a full CI run. :)

File: docs/changes/cosmology/12355.feature.rst
Comment (pllim) [reply docs/changes/cosmology/12355.feature.rst:1]: GitHub status been iffy past 2 days. You might have to wait it out.

**Review by nstarman — COMMENTED** [review]
I think just one last change then this is good to go! 🎉 
The commits should also be squashed down to a few atomic commits (https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-commits/squashing-commits-in-github-desktop).
Also, I think this PR merits a `whatsnew` entry (see https://github.com/astropy/astropy/blob/main/docs/whatsnew/6.0.rst)!

**Comment by Mubin17** [comment]
@nstarman Does the merge conflict arise because my [branch](https://github.com/Mubin17/astropy) is 291 commits behind astropy:main? If yes, should I do sync_fork?

**Comment by nstarman** [comment]
Please rebase to fix merge conflicts. We try to keep a linear history.

File: docs/whatsnew/6.0.rst
Comment (nstarman) [inline docs/whatsnew/6.0.rst:57]:
```suggestion
The :class:`~astropy.cosmology.Cosmology` class in :mod:`~astropy.cosmology` now supports the latex format in its
```

File: docs/whatsnew/6.0.rst
Comment (nstarman) [inline docs/whatsnew/6.0.rst:58]:
```suggestion
:attr:`~astropy.cosmology.Cosmology.write()` method, allowing users to export a cosmology object to a LaTeX table.::
```

**Review by nstarman — COMMENTED** [review]
Hopefully this helps with the ReadTheDocs CI.

**Review by nstarman — COMMENTED** [review]
Hopefully this helps with the ReadTheDocs CI.

**Review by nstarman — APPROVED** [review]
Thanks @Mubin17 for contributing this PR! This looks good to go!
One last bit of house-cleaning is to squash commits. If you want to do that, great, otherwise I can squash down to a few commits.
Thanks again @Mubin17. If you're interested in contributing the LaTeX reader, I'd be happy to work with you.

**Comment by Mubin17** [comment]
Thanks! @nstarman for all the help. 
I am not comfortable with squashing commits, so if you can do that it would be great!
Yes, I would love to contribute to the LaTex reader. I will open a new PR soon. 
Thanks again.

**Comment by pllim** [comment]
I already fear cosmology, but it is somehow even scarier in LaTeX! 😸 

Thanks, all!

**Pull request merged**
